### PR TITLE
Random char gen fixes

### LINKED ISSUE #16640: Random Char gen doesn't entirely respect scenarios
If you use the random character gen option and sift through the options using random scenario and then re-rolling,, You'll notice that it doesn't always seem to respect profession limitations. You'll get things like an Otaku experiment, or a bottom of the mine farmer.
You'll also notice that scenarios with special traits don't seem to register; Experiment scenario will never generate with mutations, for instance.

Pretty minor issue, but something that'd be neat to fix someday.

### PR DESCRIPTION
**Pull request by sethsimon**
Fixes #16640.
- Use the same skill costs that everywhere else uses. Increasing a skill from 0 to 2 now costs 2 points, not 1.
- Don't give the player a profession that isn't permitted by the scenario.
- Always give the player traits that are mandated by the scenario or profession.
- Never give the player traits that are forbidden by the scenario (This doesn't affect anything since currently only some challenge scenarios have forbidden traits)
- Consider all options (e.g. padded feet for the Experiment scenario) when adding traits.
- Always randomize the starting location ( see the second commit ).

### CODE REVIEW

**Comment by kilozombie** [comment]
pretty sure the first level of skills is meant to be cheaper for balance purposes.

**Comment by sethsimon** [comment]
>pretty sure the first level of skills is meant to be cheaper for balance purposes.

That was changed in #16440 (2e7fdfbe7e07a7c9f25d590c28f048e84e34958c). From reading the PR, it looks like it was an oversight during refactoring.
@Bevapdin Tagged in case you want to comment on this.

**Comment by Coolthulhu** [comment]
That first/second level thing was good for balance and I don't recall there being an explicit decision to get rid of it.
Still, random char gen should be consistent with the regular char gen. It's better if both are bugged in the same way than if random offers advantages over non-random.

**Comment by sethsimon** [comment]
In that case, I'll restore that to how it was. Also, currently profession skills are boosted when entering the description tab. So a Traceur (who starts with 4 dodging) can increase dodging from 4->5 for the same cost as increasing it from 0->1. That's not intended, is it?

**Comment by BevapDin** [comment]
> That was changed in #16440 ( 2e7fdfb ). From reading the PR, it looks like it was an oversight during refactoring.

Maybe. Probably. My personal memories don't go that far back, it might have been an oversight.

> Still, random char gen should be consistent with the regular char gen. It's better if both are bugged in the same way than if random offers advantages over non-random.

Agreed, consistence is the important thing.

File: src/newcharacter.cpp [hunk 1]
```diff
@@ -245,40 +246,42 @@ void player::randomize( const bool random_scenario, points_left &points )
...
-    }
 
     int num_gtraits = 0, num_btraits = 0, tries = 0;
     std::string rn = "";
+    add_traits(); // adds mandatory prof/scen traits.
+    for( const auto &mut : my_mutations ) {
+        if( mutation_branch::get( mut.first ).profession ) {
+            continue;
+        }
+        const mutation_branch &mut_info = mutation_branch::get( mut.first );
```
Comment (BevapDin) [inline src/newcharacter.cpp]: If you move that a few lines up (right before the `if`), you can use `mut_info` in the condition and don't need to call it twice.

File: src/newcharacter.cpp [hunk 2]
```diff
@@ -245,40 +246,42 @@ void player::randomize( const bool random_scenario, points_left &points )
...
+        if( mut_info.points >= 0 ) {
+            num_gtraits += mut_info.points;
+            points.trait_points -= mut_info.points;
+        } else {
+            num_btraits -= mut_info.points;
+            points.trait_points -= mut_info.points;
```
Comment (BevapDin) [inline src/newcharacter.cpp]: This statement seems to be the same in both branches. Therefor it should appear only once, either before or after the whole `if` block.

**Pull request merged**